### PR TITLE
Added support to read metadata fields from the pdf.

### PR DESCRIPTION
**Pull request by timKraeuter**
Added UTF-16 decoding function, getters for the metadata fields and test cases.

### CODE REVIEW

File: src/api/PDFDocument.ts [hunk 1]
```diff
@@ -210,6 +210,112 @@ export default class PDFDocument {
...
 
+  /**
+   * Get this document's title metadata as a string. The title appears in the
+   * "Document Properties" section of most PDF readers.
+   */
+  getTitle(): string {
+    const titleObject = this.getInfoDict().get(PDFName.of('Title'));
+    if (titleObject) {
+      const hexValue: string = titleObject.toString();
+      return PDFHexString.toText(this.extractFromBrackets(hexValue));
```
[the remark this answers is not in the document]
Comment (timKraeuter) [reply src/api/PDFDocument.ts]: Sounds good to me. I will change that.

File: src/core/objects/PDFHexString.ts [hunk 2]
```diff
@@ -20,6 +21,20 @@ class PDFHexString extends PDFObject {
     return new PDFHexString(hex);
   };
 
+  static toText = (value: string) => {
+    const bytes: number[] = [];
+    let i = 0;
+    while (i + 2 <= value.length) {
```
[the remark this answers is not in the document]
Comment (timKraeuter) [reply src/core/objects/PDFHexString.ts]: You are right i will change that.

[the remark this answers is not in the document]
Comment (timKraeuter) [reply src/core/objects/PDFHexString.ts]: I have fixed this now.

**Comment by timKraeuter** [comment]
I will have a look at the other 3 ways to represent a string of text in the metadata dictionaries and fix the small changes you have suggested soon.
I'm wondering how we can test the other ways without implementing them for the setters as well. I did not have a look at your integration test suite but maybe it is possible to generate some pdfs with different metadata representations and test them in the integration tests.

File: src/api/PDFDocument.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hopding) [reply src/api/PDFDocument.ts]: I’d suggest doing an `instanceof` check:
```js
if (titleObject instanceof PDFHexString || titleObject instanceof PDFString) {
  return titleObject.decodeText();
} else {
  throw new Error();
}
```

**Comment by Hopding** [comment]
I'd suggest extracting metadata from the PDFs in [`assets/pdfs`](https://github.com/Hopding/pdf-lib/tree/master/assets/pdfs) and seeing if any of them use the alternative formats. Otherwise if we can't find any sample PDFs to extract from, we'll have to add custom metadata in the tests (probably by bypassing the setters and doing it manually).

**Comment by timKraeuter** [comment]
I have tried to include your suggestions. I will have a look at the other encodings now.

PDFDocEncoding seems to be iso latin 1/ ISO 8859-1 which is as Wikipedia puts it "basis for most popular 8-bit character sets and the first block of characters in Unicode". So I think supporting it should not be too hard.

**Comment by Hopding** [comment]
@timKraeuter Just wanted to check up on this and see if you've run into issues or anything? (And no worries if you haven't had time to work on it!)

**Comment by timKraeuter** [comment]
I'm about to finish my university degree that's why I have not had any time to work on this. February is still pretty busy. I guess i can work on it beginning of march. This is still on my agenda :).

I will let you know if I need any help. Thanks ;)

**Comment by Hopding** [comment]
@timKraeuter Sounds good! No rush. Just wanted to make sure you hadn't gotten stuck or anything. And I definitely understand your position. I just finished my own degree about a month ago.

**Comment by timKraeuter** [comment]
I have implemented the decoding of the PDFDocEncoding. So the current status is the following:

String Type | Encoding
-- | --
Hex String | PDFDocEncoding :heavy_check_mark:
Hex String | UTF-16BE :heavy_check_mark:
Literal String | PDFDocEncoding
Literal String | UTF-16BE

I have committed a new test pdf for the first two cases!.

I'm wondering about the last two cases. I have seen literal Strings "(something)" but I wasn't able to generate literal strings with some special chars for PDFDocEncoding and Unicode.

How does "你怎么敢" look as a literal String encoded with UTF-16BE?
How does "˘•€" look as a literal String encoded with the PDFDocEncoding?

**Comment by kozi** [comment]
I would be very happy if this request could be accepted?

**Comment by Hopding** [comment]
@timKraeuter Sorry for the delayed response, I've been very busy at work. I'm working on getting some answers to your questions.

**Comment by timKraeuter** [comment]
No problem take your time.

**Comment by Hopding** [comment]
@timKraeuter Here's a little script you can use to figure out what a given piece of unicode text looks like when encoded as a PDF hex and literal strings:

```js
import { toHexStringOfMinLength, utf16Encode } from 'pdf-lib';

const toOctalStringOfMinLength = (num: number, minLength: number) =>
  num.toString(8).padStart(minLength, '0');

const utf16EncodeAsBytes = (input: string, byteOrderMark = true) => {
  const utf16 = utf16Encode(input, byteOrderMark);
  const bytes = new Uint8Array(utf16.length * 2);

  let offset = 0;
  for (let idx = 0, len = utf16.length; idx < len; idx++) {
    bytes[offset++] = (utf16[idx] & 0xff00) >> 8;
    bytes[offset++] = (utf16[idx] & 0x00ff) >> 0;
  }

  return bytes;
};

const printHexString = (input: string) => {
  const words = utf16Encode(input);
  const hex = Array.from(words).map((word) => toHexStringOfMinLength(word, 4));
  const hexString = '<' + hex.join('') + '>';
  console.log(hexString);
};

const printLiteralString = (input: string) => {
  const bytes = utf16EncodeAsBytes(input);
  const octal = Array.from(bytes).map((byte) => toOctalStringOfMinLength(byte, 3));
  const literalString = '(' + octal.map((byte) => `\\${byte}`).join('') + ')';
  console.log(literalString);
};

console.log('🥚 The Life of an Egg 🍳');
printHexString('🥚 The Life of an Egg 🍳');
printLiteralString('🥚 The Life of an Egg 🍳');
```

Running this script produces the following output:
```
🥚 The Life of an Egg 🍳
<FEFFD83EDD5A00200054006800650020004C0069006600650020006F006600200061006E00200045006700670020D83CDF73>
(\376\377\330\076\335\132\000\040\000\124\000\150\000\145\000\040\000\114\000\151\000\146\000\145\000\040\000\157\000\146\000\040\000\141\000\156\000\040\000\105\000\147\000\147\000\040\330\074\337\163)
```

We can of course substitute the input string with the strings you provided:
```
你怎么敢
<FEFF4F60600E4E486562>
(\376\377\117\140\140\016\116\110\145\142)

˘•€
<FEFF02D8202220AC>
(\376\377\002\330\040\042\040\254)
```

The key thing to recognize here is that in order to represent arbitrary text with literal strings, you have to use the escape sequence (the backslash syntax) defined by the PDF spec. Each escape sequence can contain 1-3 octal digits. This, of course, differs a bit from hex strings where data is encoded as hexadecimal digits rather than octal.

Hopefully this is what you were looking for. Please let me know if you have any additional questions!

**Comment by timKraeuter** [comment]
Thanks a lot. This does look exactly like what I needed. I have not thought about masking.
I'll try to work on it on monday.

**Comment by timKraeuter** [comment]
The functionality is finished now. You can have a look at the commits and we can do some cosmetics :)
I'm not fully aware of all typescript best practices i think <.<

**Comment by Hopding** [comment]
@timKraeuter This looks great! Thank you very much for all your work on this. 

Would you mind changing the target branch from `Hopding:master` to [`Hopding:ReadMetadataFields`](https://github.com/Hopding/pdf-lib/tree/ReadMetadataFields)? I'd like to tweak a couple things prior to merging it into `master`. Since it's just a number of little things it'll be easier for me to just make the changes rather than going back and forth through review comments.

Once I make those changes (hopefully this weekend at latest) and add some examples I'll cut a release with this new feature!

**Comment by timKraeuter** [comment]
I changed the target. I am looking forward to using the new release in my app with the new features.

**Comment by Hopding** [comment]
Thanks @timKraeuter! I'll try to remember to drop a comment here once I merge the changes into `master` and cut a release.

**Comment by timKraeuter** [comment]
Thanks for merging.

I have not added myself to the contributors array. How can I do that?

**Comment by Hopding** [comment]
No worries, I just added you: https://github.com/Hopding/pdf-lib/commit/972c5ebd33c029b339fe8a8c092b4e12798106fe. I'm just about ready to merge the changes into `master`. I think the only thing left is to update the README with some examples.

**Comment by timKraeuter** [comment]
Looking good thank you. You added some important test cases and simplified the code! Nice work, keep it up ;)

**Comment by Hopding** [comment]
Just merged the changes to `master`. I've got one more PR to merge and then I'm planning to cut a release. Hopefully by tomorrow evening.

**Comment by Hopding** [comment]
Version [`1.5.0`](https://www.npmjs.com/package/pdf-lib/v/1.5.0) is now published. It contains the new metadata reading feature! The full release notes are available [here](https://github.com/Hopding/pdf-lib/releases/tag/v1.5.0).

You can install this new version with npm:

```
npm install pdf-lib@1.5.0
```

It's also available on unpkg:

- https://unpkg.com/pdf-lib@1.5.0/dist/pdf-lib.js
- https://unpkg.com/pdf-lib@1.5.0/dist/pdf-lib.min.js

As well as jsDelivr:

- https://cdn.jsdelivr.net/npm/pdf-lib@1.5.0/dist/pdf-lib.js
- https://cdn.jsdelivr.net/npm/pdf-lib@1.5.0/dist/pdf-lib.min.js

**Pull request merged**